tests/test_salcommander_minimal_controllers.py

```diff
@@ -0,0 +1,147 @@
+
+    def test_salobj_controller(self) -> None:
+        for language in Language:
+            with self.subTest(language=language):
+                self.check_minimal_controller(language, "minimal_salobj_controller.py")
+
+    def check_minimal_controller(self, language: Language, exec_name: str) -> None:
+        # Start the Commander before the Controller subprocess
+        # to be sure the Commander sees telemetry from the Controller
+        # (telemetry is volatile, so has no historical data).
+        commander, initial_log = self.get_langauge_variables(language)
+        commander_path = pathlib.Path(__file__).home() / "repos/ts_sal/bin" / commander
+        assert commander_path.is_file()
+        controller_path = (
+            pathlib.Path(__file__).parent.absolute() / "controllers" / exec_name
+        )
+        assert controller_path.is_file()
+        print(f"{language.name} Commander: start {commander_path} with index={self.index} in a subprocess")
+        print(f"{language.name} Controller: start {controller_path} with index={self.index} in a subprocess")
+        with subprocess.Popen(
+            [
+                commander_path,
+                str(self.index),
+                str(INITIAL_LOG_LEVEL),
+            ],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+        ) as commander_process, subprocess.Popen(
+            [
+                str(controller_path),
+                str(self.index),
+                str(INITIAL_LOG_LEVEL)
+            ],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+        ) as controller_process:
+            print(f"{language.name} Commander: wait for completion")
+            commander_process.wait()
+            commander_output = commander_process.communicate()[0]
+            commander_output = commander_output.decode("utf-8")
+            assert commander_process.returncode == 0
+            print(f"{language.name} Controller: wait for completion")
+            controller_process.wait()
+            controller_output = controller_process.communicate()[0]
+            controller_output = controller_output.decode("utf-8")
+            print(f"{language.name} Commander: read initial logLevel")
+            self.assertIn(
+                str(initial_log) + str(INITIAL_LOG_LEVEL), commander_output
+            )
+
+            for level in (10, 52, 0):
+                print(f"{language.name} Commander: send setLogLevel(level={level}) command")
+                self.assertIn(
+                    f"Commmander:  send setLogLevel(level={level}) command",
+                    commander_output,
+                )
+                print(f"{language.name} Controller: write logLevel={level} event")
+                self.assertIn(
+                    f"Controller: writing logLevel={level} event", controller_output
+                )
+
+        print("Ensure Commander and Controller processes are terminated")
```

Comment:
If you like you can drop this line and the lines below. Though I admit it's kind of nice to have a warning that the process didn't exit as expected.

Question: do you see any of these messages when you run? I'm just wondering if the processes have reliably exited by now (if not, you might need to give them a bit of time, e.g. with a sleep).

Reply:
I do see the messages, when I run with different pytest flags (-rP instead of -ra).  I left them in as a warm and fuzzy.